lib/Sema/ConstraintSystem.cpp

```diff
@@ -2644,6 +2647,11 @@ bool ConstraintSystem::isPartialApplication(ConstraintLocator *locator) {
       locator->findFirst<LocatorPathElt::ImplicitConversion>())
     return false;
 
+  if (locator->directlyAt<OverloadedDeclRefExpr>()) {
+    auto *ODRE = castToExpr<OverloadedDeclRefExpr>(locator->getAnchor());
+    return ODRE->getFunctionRefKind() == FunctionRefKind::Unapplied;
+  }
```

Comment:
Should this also apply to regular `DeclRefExpr`s? Although `adjustFunctionTypeForConcurrency` already has `numApplies`, I wonder if `isPartialApplication` in this case should really be `numApplies < decl->getNumCurryLevels()` since `numApplies` should already account for currying? I should also note that it looks like the current logic here doesn't account for cases such as:

```swift
struct S {
  func foo() {}
}

func bar(_ x: @Sendable () -> Void) {}

let fn = S.foo(S())
bar(fn)
```

Reply:
Yeah, that sounds simpler, I'll give it a try!

Reply:
Done! We had function ref kind set incorrectly in a few places for synthesized code.